Re-order overloads and get rid of modules in TypeScript types.

arielshaqed opened the pull request:
Overloads should generally be ordered in *increasing* number of
optional args: the first matching overload is (always) selected.

Also get rid of the (single) "hot-shots" module, it needlessly
complicates type definitions.

Commit:
Re-order overloads and get rid of modules in TypeScript types.

Overloads should generally be ordered in *increasing* number of
optional args: the first matching overload is (always) selected.

Also get rid of the (single) "hot-shots" module, it needlessly
complicates type definitions.

Commit:
Ready to npm publish.

arielshaqed commented:
Tested by compiling nodeutils with it.

vogre commented on types.d.ts:
```diff
@@ -1,134 +1,132 @@
 
-declare const StatsDClient: StatsD;
-export default StatsDClient;
+export default StatsD;
```
The default exports seems incorrect before and is incorrect now - if someone imports `import StatsD from '@binaris/hot-shots'` runtime errors will follow (since there's no `exports.default`).

I think that to express `require('@binaris/hot-shots')` returns the StatsD constructor it's possible to use `export = StatsD`, but I'm not sure if it plays nice with the exported class and just removing the default export should be safer.

arielshaqed replied:
After talking F2F I see you're correct: there is no way of supporting _both_ `export = ...` _and_ new-style exports in the same file.  [TypeScript manual] says:
> Default exports are meant to act as a replacement for this behavior; however, the two are incompatible. TypeScript supports `export =` to model the traditional CommonJS and AMD workflow.

It doesn't work now and it never worked before.  I continue to `export default` so you'll be able to say
```ts
import * as sd from 'hot-shots';
const client = new sd.StatsD();
```
(which seems to work).  WDYT?

vogre replied:
`import * as sd from 'hot-shots';` works in any case, the default can still be omitted.

I have a suspicion that's is possible to make the export somehow, e.g. in Express (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express/index.d.ts) it's possible to both have
`import * as express from 'express';` and `import { Router } from 'express';`

It can't be copied directly since express is a function and StatsD a constructor, but something like

```typescript
namespace e {

  class StatsD {
    static StatsD: typeof StatsD;
  }
}
declare interface e {
  new(options?: e.ClientOptions): e.StatsD;
}

export = e;
```
should have worked, but doesn't 😞

vogre commented on types.d.ts:
```diff
@@ -1,134 +1,132 @@
-    K2 extends keyof T,
-    K3 extends keyof T,
-    K4 extends keyof T,
+export type Tags = { [key: string]: string } | string[];
```
Removing `declare module` _seems_ harmless as the main reason typescript suggest using it is to describe multiple modules in a single `.d.ts` file.

arielshaqed replied:
Ack.

vogre commented on package.json:
```diff
@@ -34,7 +35,9 @@
   },
-  "dependencies": {},
+  "dependencies": {
+    "@types/node": "^11.10.4"
```
I think this dependency can be dropped (no need forcing a node version).

arielshaqed replied:
Done.

vogre reviewed (changes requested):
Thanks!

One major request about the export default.

arielshaqed reviewed (commented):
Thanks!  PTAL.

Commit:
Don't rely on @types/node.

It would be a "peer dev dependency if peer is using TypeScript";
that's impossible to express in `package.json`.
